junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/NamespaceAwareStore.java

```diff
@@ -41,49 +41,58 @@ public NamespaceAwareStore(NamespacedHierarchicalStore<Namespace> valuesStore, N
 	@Override
 	public @Nullable Object get(Object key) {
 		Preconditions.notNull(key, "key must not be null");
-		return accessStore(() -> this.valuesStore.get(this.namespace, key));
+		Supplier<@Nullable Object> action = () -> this.valuesStore.get(this.namespace, key);
+		return accessStore(action);
 	}
 
 	@Override
 	public <T> @Nullable T get(Object key, Class<T> requiredType) {
 		Preconditions.notNull(key, "key must not be null");
 		Preconditions.notNull(requiredType, "requiredType must not be null");
-		return accessStore(() -> this.valuesStore.get(this.namespace, key, requiredType));
+		Supplier<@Nullable T> action = () -> this.valuesStore.get(this.namespace, key, requiredType);
+		return accessStore(action);
 	}
 
 	@Override
-	public <K, V> @Nullable Object getOrComputeIfAbsent(K key, Function<K, V> defaultCreator) {
+	public <K, V extends @Nullable Object> @Nullable Object getOrComputeIfAbsent(K key, Function<K, V> defaultCreator) {
 		Preconditions.notNull(key, "key must not be null");
 		Preconditions.notNull(defaultCreator, "defaultCreator function must not be null");
-		return accessStore(() -> this.valuesStore.getOrComputeIfAbsent(this.namespace, key, defaultCreator));
+		Supplier<@Nullable Object> action = () -> this.valuesStore.getOrComputeIfAbsent(this.namespace, key,
+			defaultCreator);
+		return accessStore(action);
 	}
 
 	@Override
-	public <K, V> @Nullable V getOrComputeIfAbsent(K key, Function<K, V> defaultCreator, Class<V> requiredType) {
+	public <K, V extends @Nullable Object> @Nullable V getOrComputeIfAbsent(K key, Function<K, V> defaultCreator,
```

Comment:
Should this be `Function<? super K, ? extends V>` ?

Reply:
Why the return value is "forced nullable"? I guess the nullability should be the same as V's one

Reply:
Javadoc says the key must not be `null`: https://github.com/junit-team/junit5/blob/7d02a6a2521617ab8502f8614794b7bb0a568603/junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/ExtensionContext.java#L672

Reply:
If class-level "key-value" parameter is not feasible, then it should be very well feasible to add type parameter to the key:

```java
interface Store {
  <KeyType, ValueType> ValueType get(StoreKey<KeyType, ValueType> key);
  <KeyType, ValueType> ValueType put(StoreKey<KeyType, ValueType> key, ValueType value);
}

interface StoreKey<KeyType, ValueType> {
}
```

Then the compiler would be able to type-check the values associated with every key